### ISSUE TITLE
Slow loading of map tiles in 6.0.2

### ISSUE THREAD

**Issue by s-ilnar** [osmdroid/osmdroid#1193]
I use version 6.0.2. Compared to version 6.0.1, version 6.0.2 takes a long time to download map tiles, especially noticeable on Android 8.  
I use  
`Configuration.getInstance().setUserAgentValue(getContext().getPackageName());`
`mapView.setTileSource(TileSourceFactory.MAPNIK);`

` implementation group: 'org.osmdroid', name: 'osmdroid-android', version: '6.0.2'`

**Comment by monsieurtanuki** [comment]
@s-ilnar Could it be because we switched Mapnik tile source to `https` when fixing #1050 for 6.0.2?
Could you try to create your own `TileSource` as a "http Mapnik" and check if it is faster? Something like
````java
public static final OnlineTileSourceBase HTTP_MAPNIK = new XYTileSource("HttpMapnik",
	0, 19, 256, ".png", new String[] {
		"http://a.tile.openstreetmap.org/",
		"http://b.tile.openstreetmap.org/",
		"http://c.tile.openstreetmap.org/" },
	"© OpenStreetMap contributors");
````

**Comment by s-ilnar** [comment]
Hello, thanks for the quick response! Unfortunately, I could not quickly check. Just tried your option, didn't help. On version 6.0.2 on Android 8.0-8.1 long drawn map tiles

**Comment by monsieurtanuki** [comment]
@s-ilnar What is slower? The download or the display?
How slower is it compared to 6.0.1? Can you give metrics?
On which tiles? Can you give a range of tiles which are slower to download and/or display in 6.0.2?

**Comment by s-ilnar** [comment]
On version 6.0.2 if you enable caching then slightly improved, but if you switch to version 6.0.1 it becomes even better. I guess it takes a long time to drawing. How can I check more precisely?

**Comment by monsieurtanuki** [comment]
I assume that if you switch the debug mode on (probably having to do with `Configuration`), you'll get stats in the logs.
Then, starting from the same point (e.g. empy database, same tiles to download and to display, same map orientation - 0 would make sense), collect statistics and find what is slower (e.g. the display of each map is x% slower - from y ms to z ms).

**Comment by monsieurtanuki** [comment]
@s-ilnar Actually a new release has just been released: 6.0.3. Perhaps it could solve your performance issue.

**Comment by chrisss404** [comment]
I enabled all debug flags I could find:

    Configuration.getInstance().setDebugMode(true);
    Configuration.getInstance().setDebugMapView(true);
    Configuration.getInstance().setDebugMapTileDownloader(true);
    Configuration.getInstance().setDebugTileProviders(true);

And here are my results:

6.0.1 A - 2018-12-12 18:07:50.133 - 2018-12-12 18:07:53.864 => 3.731 - [version-6.0.1-A.log](https://github.com/osmdroid/osmdroid/files/2672937/version-6.0.1-A.log)
6.0.3 A - 2018-12-12 18:10:53.844 - 2018-12-12 18:10:58.180 => 4.336 - [version-6.0.3-A.log](https://github.com/osmdroid/osmdroid/files/2672940/version-6.0.3-A.log)

6.0.1 B - 2018-12-12 18:31:27.624 - 2018-12-12 18:31:30.985 => 3.361 - [version-6.0.1-B.log](https://github.com/osmdroid/osmdroid/files/2672939/version-6.0.1-B.log)
6.0.3 B - 2018-12-12 18:33:33.476 - 2018-12-12 18:33:37.515 => 4.039 - [version-6.0.3-B.log](https://github.com/osmdroid/osmdroid/files/2672942/version-6.0.3-B.log)

It seems that version 6.0.1 is downloading less map tiles than 6.0.3 for the same map section.

**Comment by monsieurtanuki** [comment]
@chrisss404 Thank you for your logs and your comment!

I'd need time to find relevant data in the 1000s of lines of your logs, but maybe it's not necessary as your comment may lead to the answer.

Recently (I lost track of the version numbers) was introduced the concepts of tile pre-cache and tile cache protection.
Short version: it's about pre-caching some tiles (e.g. pre-downloading them and putting them in the memory cache) and protecting them from being removed from the memory cache. Those tiles are typically the tiles in the border of the displayed tiles, and the same tiles but on zooms +1 and -1.
The initial idea was to have a smooth display when changing zoom level or panning the map, as the expected tiles are already in the cache. The downside is that there are more downloads than strictly needed at time t, and perhaps that is what causes overall slower download impression.

It's possible for you to change this default behavior as a test, and probably the best idea would be to change the default behavior altogether, but I'd like you to test first.
What you need to add is something like:
````java
mapView.getTileCache().getProtectedTileComputers().clear();
mapView.getTileCache().setAutoEnsureCapacity(false);
````
Could you test if you manage to get the same download speed impression with those lines in 6.0.3 as with version 6.0.1?

PS: "there are more downloads than strictly needed" - for the record it's not only about pre-_downloading_ tiles, tiles can be found in other providers like `SqlTileWriter`, which is less costly, resource-wise

**Comment by chrisss404** [comment]
Added the following

    map.getTileProvider().getTileCache().getProtectedTileComputers().clear();
    map.getTileProvider().getTileCache().setAutoEnsureCapacity(false);

And my results

6.0.1 A - 2018-12-12 21:24:03.420 - 2018-12-12 21:24:07.628 => 4.208 - [version-6.0.1-A.log](https://github.com/osmdroid/osmdroid/files/2673449/version-6.0.1-A.log)
6.0.3 A - 2018-12-12 21:26:49.256 - 2018-12-12 21:26:53.558 => 4.302 - [version-6.0.3-A.log](https://github.com/osmdroid/osmdroid/files/2673451/version-6.0.3-A.log)

6.0.1 B - 2018-12-12 21:31:16.406 - 2018-12-12 21:31:20.059 => 3.653 - [version-6.0.1-B.log](https://github.com/osmdroid/osmdroid/files/2673450/version-6.0.1-B.log)
6.0.3 B - 2018-12-12 21:33:11.642 - 2018-12-12 21:33:16.010 => 4.368 - [version-6.0.3-B.log](https://github.com/osmdroid/osmdroid/files/2673452/version-6.0.3-B.log)

I am not sure if this result can be considered as an improvement but it is definitely no worse than before.

**Comment by monsieurtanuki** [comment]
@chrisss404 I'm not sure I get what you mean with "no worse than _before_".
Do you mean that my suggestion of 6.0.3 fix makes your map display as fast as 6.0.1? Or as fast as 6.0.3 without fix (which would mean the fix is not a fix).
I wasn't expecting to have better performances than in 6.0.1 with that 6.0.3 fix, just going back to your 6.0.1 performance.
Tell me if you think the 6.0.3 fix is relevant and I'll integrate it, and then the issue can be closed.

**Comment by chrisss404** [comment]
I started using OsmDroid on Monday, therefore it is very difficult for me to answer your questions.

The numbers show that there is no significant improvement when applying your suggested fix.

Regarding perceived performance, see following recordings: [6.0.01](https://www.dropbox.com/s/ym2f451gfeo1u4e/version-6.0.1.mp4?dl=0), [6.0.3-with-fix](https://www.dropbox.com/s/kp1ugvzqlrex1w3/version-6.0.3-with-fix.mp4?dl=0), and [6.0.3-without-fix](https://www.dropbox.com/s/2gnem9vbmm4933l/version-6.0.3-without-fix.mp4?dl=0).

In conclusion: it seems that the changes added with version 6.0.2 have introduced a performance impact. Downloading and displaying a certain map section takes about 20% longer than before.

Sorry that I am not more helpful.

**Comment by monsieurtanuki** [comment]
OK @chrisss404, thanks to your videos I've finally understood what you meant.

Cf. #1004: we recently implemented the "offline first" strategy, which is _very_ relevant if you have a slow connection - in the meanwhile you can see something more interesting than a gray square.
But it looks like in your case (or at least in those tests with a decent connection), the tile approximations take a time than you consider as slower performances.

In order to validate my hypothesis, I suggest that you:
1. copy `MapTileProviderBasic` into your own `MyOwnMapTileProviderBasic`
1. inside your `MyOwnMapTileProviderBasic` class, move the "approximationProvider" block _after_ the "downloaderProvider" block
1. have your `MapView` use your `MyOwnMapTileProviderBasic`
1. again, use the following code
````java
map.getTileProvider().getTileCache().getProtectedTileComputers().clear();
map.getTileProvider().getTileCache().setAutoEnsureCapacity(false);
````
5. check if you have the same behavior as in 6.0.1

**Comment by chrisss404** [comment]
@monsieurtanuki thank you for your detailed instructions. 

I can confirm that I'm getting the same behavior as with 6.0.1 using the adapted map tile provider. See [6.3.0-with-adapted-tile-provider](https://www.dropbox.com/s/u4q8fjctb0bhivr/version-6.3.0-with-adapted-tile-provider.mp4?dl=0).

Just for reference, here are my changes: [GitHubIssueMapTileProviderBasic.txt](https://github.com/osmdroid/osmdroid/files/2683813/GitHubIssueMapTileProviderBasic.txt)

    map.setTileProvider(new GitHubIssueMapTileProviderBasic(this));
    map.getTileProvider().getTileCache().getProtectedTileComputers().clear();
    map.getTileProvider().getTileCache().setAutoEnsureCapacity(false);

Thanks again and best regards

**Comment by monsieurtanuki** [comment]
@chrisss404 I've just PR'ed #1235, which includes a `MapTileProviderBasic.setOfflineFirst(boolean)` method.
In your case you need to call it with `false`. And you won't need your `GitHubIssueMapTileProviderBasic` class anymore.
Tell me if it works.

@s-ilnar Feel free to test and close the issue if relevant.
